### PR TITLE
[gguf] get optional `paramCount`

### PR DESCRIPTION
**Pull request by mishig25**
Optionally get `paramCount` from parsing the gguf file so that we can use this information in hub-backend

### CODE REVIEW

File: packages/gguf/src/gguf.ts [hunk 1]
```diff
@@ -293,5 +295,41 @@ export async function gguf(
...
+				if (shardIdx !== shard) {
+					urls.push(`${prefix}-${shardIdx.toString().padStart(5, "0")}-of-${total.toString().padStart(5, "0")}.gguf`);
+				}
+			}
+
+			const PARALLEL_DOWNLOADS = 20;
+			const shardsParamCount = await promisesQueue(
+				urls.map((shardUrl) => async () => {
+					const { tensorInfos: shardTesnorInfos } = await gguf(shardUrl);
+					const shardParamCount = shardTesnorInfos
```
[the remark this answers is not in the document]
Comment (mishig25) [reply packages/gguf/src/gguf.ts]: handled in 6be0dc120973a5942847cd0ee90bae1da468b01b

**Comment by ngxson** [comment]
I additionally reviewed this PR. It's LGTM overall.

One small suggestion is that maybe we should abstract/generalize the support for shared gguf (while using the correct metadata) in the future. If you're OK with this idea, I'll have a look on it.

File: packages/gguf/src/gguf.ts [hunk 2]
```diff
@@ -183,6 +184,7 @@ export async function gguf(
 		 * Custom fetch function to use instead of the default one, for example to use a proxy or edit headers.
 		 */
 		fetch?: typeof fetch;
+		calculateParamCount?: boolean;
```
Comment (julien-c) [inline packages/gguf/src/gguf.ts:187]: why not the same parameter name as for safetensors (`computeParametersCount`)?

File: packages/gguf/src/types.ts [hunk 3]
```diff
@@ -138,4 +138,5 @@ export interface GGUFTensorInfo {
 export interface GGUFParseOutput {
 	metadata: GGUFMetadata;
 	tensorInfos: GGUFTensorInfo[];
+	paramCount?: number;
```
Comment (julien-c) [inline packages/gguf/src/types.ts:141]: why not the same prop name as for safetensors (`parameterCount`)? Besides i don't think this should be in this type (which is strictly what is in GGUF), but just in the function return type

File: packages/gguf/src/gguf.ts [hunk 4]
```diff
@@ -293,5 +295,41 @@ export async function gguf(
 		});
 	}
 
-	return { metadata, tensorInfos };
+	let paramCount: number | undefined = undefined;
+
+	if (params?.calculateParamCount) {
+		paramCount = tensorInfos
+			.map(({ shape }) => shape.reduce((acc, val) => acc * Number(val), 1))
+			.reduce((acc, val) => acc + val, 0);
```
Comment (julien-c) [inline packages/gguf/src/gguf.ts:303]: i agree with this part, but shouldn't we return a `Record<Dtype, number>` like in safetensors? (no strong opinion, I'm ok with only returning a number, that's fine)

File: packages/gguf/src/gguf.ts [hunk 5]
```diff
@@ -293,5 +295,41 @@ export async function gguf(
...
+
+		const ggufShardFileInfo = parseGgufShardFile(url);
+		if (ggufShardFileInfo) {
```
Comment (julien-c) [inline packages/gguf/src/gguf.ts:306]:
```suggestion
		if (ggufShardFileInfo && params?.fetchAllShards) {
```

IMO this behavior (fetching other shards magically) is quite unintuitive, so I would at least add a special flag for it.

Or maybe better API-wise to split this into another method

**Review by julien-c — CHANGES_REQUESTED** [review]

**Review by julien-c — APPROVED** [review]

**Pull request merged**